### ISSUE TITLE
Detect Reagent component function calls for arity linting

### ISSUE THREAD

**Issue by borkdude** [clj-kondo/clj-kondo#25]
```
(defn my-component []
  [:div "Hello"])

[:div [my-component 1 2 3]] ;; <- arity error
```

**Comment by borkdude** [comment]
How do we detect if a function is intended as a Reagent component and hence a "vector call" should be interpreted as a function call?

A combination of one or more of these:

1) we're in .cljs or a CLJS reader conditional part in .cljc
2) We could look at the namespace requires. If reagent.core occurs there, it's extremely likely
3) Check if a function uses r/create-class

**Comment by agoldhammer** [comment]
In addition, using the idiomatic reagent  [component.classname ...]  or [component#id ...] produces "unresolved symbol" errors. Can this be fixed?

**Comment by borkdude** [comment]
@agoldhammer That's a separate issue. Can you please make an issue with a self-consistent example (.cljs file) that I can lint?

**Comment by borkdude** [comment]
@agoldhammer According to a Reagent maintainer on Clojurians slack (@Deraen) this is not supported.

**Comment by agoldhammer** [comment]
Sorry, my mistake. Supported in Hiccup, not Reagent. Please ignore.

**Comment by borkdude** [comment]
Thought about this long enough and I don't see a feasible way to do this without producing false positives.

**Comment by borkdude** [comment]
@Deraen:

Now that reagent has a special macro to define components, I think we can finally get arity warnings for component "calls" too by expanding [foo ...] into (foo ...) in the component body in a clj-kondo hook. This can be done by postwalking the body of `defc`. This would be a great addition to the new reagent release, don't you think?

**Comment by borkdude** [comment]
Some complication to this is that you shouldn't expand vectors starting with a symbol to a fn call, e.g. in `(let [foo 1])` so it's not trivial to do maybe. Maybe a good rule of thumb is not expand vectors that are part of a fn call themselves.

**Comment by Deraen** [comment]
I guess keeping a registry of all symbols created with defc, and then checking the vectors in any defn/defc or maybe all vectors where the first element is one those symbols would require multi pass check and isn't feasible?

**Comment by sirmspencer** [comment]
> I guess keeping a registry of all symbols created with defc, and then checking the vectors in any defn/defc or maybe all vectors where the first element is one those symbols would require multi pass check and isn't feasible?

I was going to suggest something similar.  I think its ok to look at all vectors in the code, even in a let binding.  If the first element is defined in malli-type-cljs/config.edn for example, then you know to treat the vector as if it were a function.  

In this example, I have a simple component with a schema attached.  If I call it as a function it knows to lint.  If I call it as a component (hiccup) it doesn't. 

<img width="1092" height="762" alt="Image" src="https://github.com/user-attachments/assets/4c16f413-cdb5-43fc-993a-7c9abb30816d" />

Somewhere in that check of the function its looking up the lint rules for -common-button in the config.edn file.  That same logic applies to the vector form.  A difference might be what to do if no value is found.  Function calls may have some basic linting that can be skipped for vectors that don't resolve with the first element as a known function.

You could make it a setting so that projects not using reagent don't waste the cpu time on checking vectors.
